File: 112_pathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if root is None:
            return False
        S = [(root, sum)]
        while S:
            (node, sum) = S[0]
            S = S[1:]
            if node.left is None and node.right is None:
                if node.val-sum == 0:
                    return True
            if node.left:
                S.append((node.left, sum-node.val))
            if node.right:
                S.append((node.right, sum-node.val))
        return False

File: 112_pathSum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Accum struct {
    Node    *TreeNode
    Sum     int
}

func hasPathSum(root *TreeNode, targetSum int) bool {
    
    if root == nil {
        return false
    }
    
    stk := make([]Accum, 0)
    stk = append(stk, Accum{root, 0})
    
    for len(stk) > 0 {
        top := stk[len(stk)-1]
        stk = stk[:len(stk)-1]

        if top.Node.Left == nil && top.Node.Right == nil {
            if top.Node.Val + top.Sum == targetSum {
                return true
            }
        } else {
            if top.Node.Left != nil {
                stk = append(stk, Accum{top.Node.Left, top.Sum + top.Node.Val})
            }
            if top.Node.Right != nil {
                stk = append(stk, Accum{top.Node.Right, top.Sum + top.Node.Val})
            }
        }  
    }
    return false
}
